README.md
# Morse audio decoder

This program is in solution to [Wunderdog Wundernut vol. 11][wundernut], whose instructions can be found in [their GitHub][wundernut-11-github].

The program reads wav audio file, and outputs decoded morse code in standard output.

## Quickstart

### Installation

#### Option 1 - pip

You can install this package from pip, with

    pip install morse-audio-decoder

#### Option 2 - Local install from sources

Clone code repository from your local machine, install from there:

    git clone https://github.com/mkouhia/morse-audio-decoder.git
    cd morse-audio-decoder
    poetry build
    # take note of the build step output, install package from the dist folder
    pip install dist/PRODUCED_PACKAGE.whl

### Usage

To run the script installed with pip, perform

    morse-audio-decoder WAVFILE

or alternatively,

    python -m morse_audio_decoder WAVFILE

where `WAVFILE` is path to the audio file to be processed.

The program decodes audio morse code in the WAVFILE argument, and writes translation to standard output.
See program help with command line flag `-h`:

    $ morse-audio-decoder -h
    usage: morse-audio-decoder [-h] WAVFILE

    Read audio file in WAV format, extract the morse code and write translated text into standard output.

    positional arguments:
    WAVFILE     Input audio file

    options:
    -h, --help  show this help message and exit

### Usage in Python

```python
from morse_audio_decoder.morse import MorseCode

morse_code = MorseCode.from_wavfile("/path/to/file.wav")
out = morse_code.decode()
print(out)
```


## Technical description

The program works in following steps

1. Read in the WAV file.
2. Extract [analytic envelope][envelope-wikipedia] from the signal by calculating moving RMS amplitude with [Hann window][hann-wikipedia] of default 0.01 second width. This envelope signal is smooth and always greater than or equal to zero.
3. Convert envelope to binary 0/1 signal by applying threshold, by default `0.5 * max(envelope)`
4. Calculate durations of continuous on/off samples
5. Identify dash/dot characters and different breaks with [K-Means clustering][kmeans-wikipedia]. The lengths of periods are compared, and then labeled automatically based on number of samples.
6. Create dash/dot character array, which is then broken to pieces by character and word space indices
7. Translate morse coded characters into plain text, print output

Exploratory data analysis and first program implementation is performed in [this jupyter notebook][initial-notebook]. The notebook is not updated; actual implementation differs.


### Restrictions

This decoder has been tested and developed with inputs that have
- no noise
- constant keying speed
- constant tone pitch
- single input channel.

If the decoder were to be extended to noisy inputs with major differences, at least following changes would be required
- pitch detection in moving time
- signal extraction with narrow bandpass filter, based on identified pitch
- keying speed detection (characters/words per minute)
- decoding in smaller time steps, taking into account speed changes.

The program is also not intended to identify single characters, as the precision will be lower with shorter inputs.

## Development

### Environment

Requirements:
- Python 3.10
- Poetry (see [installation instructions][poetry-install])

Dependencies:
- Numpy
- Scikit-learn

1. Install dependencies with `poetry install`
2. Enter environment with `poetry shell`


### Code quality and testing

All code is to be formatted with `black`:

    black **/*.py

and code quality checked with `pylint`:

    pylint **/*.py

Tests should be written in `pytest`, targeting maximum practical code coverage. Tests are run with:

    pytest

and test coverage checked with

    pytest --cov

Optionally, html test coverage reports can be produced with

    pytest --cov morse_audio_decoder --cov-report html

### Contributions

Contributions are welcome. Please place an issue or a pull request.


[wundernut]: https://www.wunderdog.fi/wundernut
[wundernut-11-github]: https://github.com/wunderdogsw/wundernut-vol11
[envelope-wikipedia]: https://en.wikipedia.org/wiki/Envelope_(waves)
[hann-wikipedia]: https://en.wikipedia.org/wiki/Hann_function
[initial-notebook]: notebooks/2022-02-23%20Wundernut%2011%20exploration.ipynb
[kmeans-wikipedia]: https://en.wikipedia.org/wiki/K-means_clustering
[poetry-install]: https://python-poetry.org/docs/#installation


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.0] - 2022-03-01
### Added
- Decode morse code in a .wav file into plain text
- Speed and pitch independent signal detection
- Command line script `morse-audio-decoder`
- Notebook for explaining line-of-thought
- Tests for all methods

[Unreleased]: https://github.com/mkouhia/morse-audio-decoder/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/mkouhia/morse-audio-decoder/releases/tag/v0.1.0

tests/test_io.py
"""Test input/output"""

from pathlib import Path

import numpy as np

from morse_audio_decoder.io import read_wave
from .common_fixtures import wav_file_fx  # pylint: disable=unused-import


def test_read_wave_frame_rate(wav_file: Path):
    """Expect 44100 Hz frame rate"""
    frame_rate, _ = read_wave(wav_file)
    assert frame_rate == 44100


def test_read_wave_shape(wav_file: Path):
    """Expect 1D data for mono audio"""
    _, data = read_wave(wav_file)
    assert len(data.shape) == 1


def test_read_wave_dtype(wav_file: Path):
    """Expect 16 bit integer array"""
    _, data = read_wave(wav_file)
    assert data.dtype == np.int16


tests/test_morse.py
"""Test morse code handling"""

from pathlib import Path

import numpy as np
from numpy.testing import assert_array_equal
import pytest

from morse_audio_decoder.morse import MorseCode

# pylint: disable=unused-import
from .common_fixtures import (
    wav_file_fx,
    wav_file_8bit_fx,
)


@pytest.fixture(name="hello_world_morse")
def hello_world_morse_fx() -> str:
    """HELLO WORLD string in morse"""
    return ".... . .-.. .-.. ---|.-- --- .-. .-.. -.."


@pytest.fixture(name="hello_data")
def hello_data_fx(hello_world_morse: str) -> np.ndarray:
    """Add dummy data for HELLO WORLD string"""
    hello_world_str = (
        "000"
        + (
            hello_world_morse.replace(" ", "00")
            .replace("|", "000000")
            .replace(".", "10")
            .replace("-", "1110")
        )
        + "00"
    )
    return np.repeat(np.array([int(i) for i in hello_world_str]), 44100 * 60 // 1000)


@pytest.fixture(name="off_samples")
def off_samples_fx(hello_world_morse: str):
    """Serve off samples for tests"""
    pairs = [
        hello_world_morse.replace("-", ".")[i : i + 2]
        for i in range(len(hello_world_morse) - 1)
    ]
    pair_off_lens = {"..": 1, ". ": 3, ".|": 7}
    expected_off = (
        np.array(
            list(map(pair_off_lens.get, [p for p in pairs if p not in [" .", "|."]]))
        )
        * 44100
        * 60
        // 1000
    )
    return expected_off


def test_from_wavfile(wav_file: Path):
    """Constructor runs and data is initalized with something"""
    received = MorseCode.from_wavfile(wav_file)

    assert len(received.data) > 0


def test_from_wavfile_8bit(wav_file_8bit: Path):
    """Construction with 8bit input data also works"""
    received = MorseCode.from_wavfile(wav_file_8bit)

    assert len(received.data) > 0


def test_decode(hello_data: np.ndarray):
    """Dummy data decoding works"""
    received = MorseCode(hello_data).decode()
    assert received == "HELLO WORLD"


def test_morse_to_char():
    """All alphanumeric characters and full stop are in values"""
    received = MorseCode(np.empty(1)).morse_to_char
    expected_chars = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ."

    assert set(received.values()).issuperset(expected_chars)


def test_morse_to_char_cached(mocker):
    """Cached dictionary is read from MorseCode._morse_to_char"""
    expected = {"..": "A"}
    mocker.patch("morse_audio_decoder.morse.MorseCode._morse_to_char", expected)
    morse = MorseCode(np.empty(1))

    received = morse.morse_to_char

    assert received == expected


def test_on_off_samples(
    hello_data: np.ndarray, hello_world_morse: str, off_samples: np.ndarray
):
    """Count of samples is detected correctly"""
    on_str = (
        hello_world_morse.replace(" ", "")
        .replace("|", "")
        .replace(".", "1")
        .replace("-", "3")
    )
    expected_on = np.array([int(i) for i in on_str]) * 44100 * 60 // 1000

    # pylint: disable=protected-access
    received_on, received_off = MorseCode(hello_data)._on_off_samples()

    assert_array_equal(received_on, expected_on)
    assert_array_equal(received_off, off_samples)


def test_on_off_samples_no_padding(hello_data: np.ndarray, hello_world_morse: str):
    """Test that output is equal, whether or not there is empty space at start/end"""
    test_str = (
        hello_world_morse.replace(" ", "00")
        .replace("|", "000000")
        .replace(".", "10")
        .replace("-", "1110")
    )
    test_str = test_str[: len(test_str) - 1]  # remove final zero

    hello_data_no_pad = np.repeat(
        np.array([int(i) for i in test_str]), 44100 * 60 // 1000
    )

    # pylint: disable=protected-access
    expected_on, expected_off = MorseCode(hello_data)._on_off_samples()
    received_on, received_off = MorseCode(hello_data_no_pad)._on_off_samples()

    assert_array_equal(received_on, expected_on)
    assert_array_equal(received_off, expected_off)


def test_dash_dot_characters(hello_world_morse: str):
    """Sample length to dash/dot conversion"""
    dash_dots = hello_world_morse.replace(" ", "").replace("|", "")

    on_str = dash_dots.replace(".", "1").replace("-", "3")
    on_samples = np.array([int(i) for i in on_str]) * 44100 * 60 // 1000

    # pylint: disable=protected-access
    received = MorseCode(np.empty(1))._dash_dot_characters(on_samples)
    expected = np.array(list(dash_dots))

    assert_array_equal(received, expected)


def test_break_spaces(hello_world_morse: str, off_samples: np.ndarray):
    """Space and word breaks are found correctly"""
    all_same_spaces = hello_world_morse.replace("|", " ")
    char_break_idx = np.nonzero(np.array(list(all_same_spaces)) == " ")[0]
    char_break_idx = char_break_idx - np.arange(len(char_break_idx))

    word_space_idx = np.nonzero(np.array(list("HELLO WORLD")) == " ")[0]

    # pylint: disable=protected-access
    received_cb, received_wb = MorseCode(np.empty(1))._break_spaces(off_samples)

    assert_array_equal(received_cb, char_break_idx)
    assert_array_equal(received_wb, word_space_idx)


def test_morse_words(hello_world_morse: str):
    """Expansion of character and space arrays works as expected"""
    only_dash_dots = hello_world_morse.replace("|", "").replace(" ", "")
    dash_dot_characters = np.array(list(only_dash_dots))

    all_same_spaces = hello_world_morse.replace("|", " ")
    char_break_idx = np.nonzero(np.array(list(all_same_spaces)) == " ")[0]
    char_break_idx = char_break_idx - np.arange(len(char_break_idx))

    word_space_idx = np.nonzero(np.array(list("HELLO WORLD")) == " ")[0]

    # pylint: disable=protected-access
    received = MorseCode(np.zeros(10))._morse_words(
        dash_dot_characters, char_break_idx, word_space_idx
    )
    expected = [word.split(" ") for word in hello_world_morse.split("|")]

    assert received == expected


def test_translate(hello_world_morse: str):
    """Correct translation is received"""
    morse_words = [word.split(" ") for word in hello_world_morse.split("|")]

    # pylint: disable=protected-access
    received = MorseCode(np.zeros(10))._translate(morse_words)
    expected = "HELLO WORLD"

    assert received == expected


tests/common_fixtures.py
"""Fixtures that are common with multiple test classes"""

from pathlib import Path

import pytest


@pytest.fixture(name="wav_file")
def wav_file_fx() -> Path:
    """Provide path to beep.wav"""
    return Path(__file__).parent / "fixtures" / "beep.wav"


morse_audio_decoder/io.py
"""Input/output"""

import os
import wave

import numpy as np


def read_wave(file: os.PathLike) -> tuple[int, np.ndarray]:
    """Read WAV file into numpy array

    NOTE: only mono audio is supported. Multi-channel audio is interlaced,
    and would need to be de-interlaced into a 2D array.

    Args:
        file (os.PathLike): input WAV file

    Returns:
        tuple[int, np.ndarray]: sample rate, data

        Data type is determined from the file; for 16bit PCM (as in competition),
        the output data type is int16. For mono audio, return shape is 1D array.
    """
    with wave.open(str(file), "rb") as wav_file:
        buffer = wav_file.readframes(wav_file.getnframes())
        sample_width_bits = wav_file.getsampwidth() * 8
        data = np.frombuffer(buffer, dtype=f"int{sample_width_bits}")
        if wav_file.getnchannels() > 1:
            raise NotImplementedError(
                "Cannot read WAV file with more than one channels, found: "
                + str(wav_file.getnchannels())
            )
        return wav_file.getframerate(), data


morse_audio_decoder/__main__.py
"""Command line interface"""

import argparse
import os
from pathlib import Path
import sys

from morse_audio_decoder.morse import MorseCode


def main(file: os.PathLike) -> None:
    """Read WAV file, process it and write outputs to stdout

    Args:
        file (os.PathLike): path to WAV file

    Raises:
        UserWarning: If dash/dot separation cannot be made unambiguosly,
            or if input file does not exist.
    """
    if not Path(file).exists():
        raise UserWarning(f"File {file} not found, exiting.")

    decoded = MorseCode.from_wavfile(file).decode()
    sys.stdout.write(decoded + "\n")


def parse_args(args: list[str]) -> argparse.Namespace:
    """Parse arguments from command line"""
    parser = argparse.ArgumentParser(
        description="""Read audio file in WAV format, extract the morse code and
        write translated text into standard output."""
    )
    parser.add_argument("WAVFILE", help="Input audio file")
    return parser.parse_args(args)


if __name__ == "__main__":
    program_arguments = parse_args(sys.argv[1:])
    try:
        main(program_arguments.WAVFILE)
    except UserWarning as err:
        sys.stderr.write(f"{err}\n")
        sys.exit(1)


morse_audio_decoder/processing.py
"""Audio signal processing"""

import numpy as np


def smoothed_power(
    data: np.ndarray, window_size: int, mode: str = "valid"
) -> np.ndarray:
    """Calculate moving time window RMS power for a signal

    Produce amplitude envelope, which reperesents signal power over time.
    Power is calculated as RMS (root mean squared) value.
    The envelope is smoothed by Hann window convolution.

    Args:
        data (np.ndarray): Input data
        window_size (int): Smoothing window length, samples
        mode (str): Convolution mode, one of  "same" and "valid".
            When "same", return same length array as in input; when "valid",
            convolution is only given for signal points that fully overlap with
            the smoothing window. See np.convolve documentation
            for further explanation.

    Returns:
        np.ndarray: smoothed array
    """
    # Convert data in order to avoid truncation errors
    if np.issubdtype(data.dtype, np.integer):
        secure_data = data.astype(np.int32) if data.itemsize < 32 else data
    else:
        secure_data = data.astype(np.float32) if data.itemsize < 32 else data

    # Create window with integral=1 -> multiplication results in weighted average
    window = np.hanning(window_size)
    window = window / sum(window)

    squared = np.power(secure_data, 2)

    return np.sqrt(np.convolve(squared, window, mode)).astype(data.dtype)


def squared_signal(data: np.ndarray, threshold: int | float = None) -> np.ndarray:
    """Convert signal to binary 0/1 based on threshold value

    Args:
        data (np.ndarray): Input data
        threshold (int | float, optional): Threshold value. All values in data
            that are smaller than threshold value are converted to 0, larger values
            are converted to 1. Defaults 0.5 * max(data).

    Returns:
        np.ndarray: Binary array of int8 dtype, same shape as original
    """
    threshold = threshold or 0.5 * np.max(data)
    return np.where(data > threshold, 1, 0).astype(np.int8)


morse_audio_decoder/morse.py
"""Morse code handling"""

from configparser import ConfigParser
import os
from pathlib import Path
import sys
import warnings

import numpy as np
import sklearn.cluster
import sklearn.exceptions

from .io import read_wave
from .processing import smoothed_power, squared_signal


class MorseCode:

    """Morse code

    Attributes:
        data (np.ndarray): 1D binary array, representing morse code in time
    """

    _morse_to_char: dict = None

    def __init__(self, data: np.ndarray, sample_rate: int = None):
        """Initialize code with binary data

        Args:
            data (np.ndarray): 1D binary array, representing morse code in time
            sample_rate (np.ndarray): Audio sampling rate. Default: None.
        """
        self.data = data
        self.sample_rate = sample_rate

    @classmethod
    def from_wavfile(cls, file: os.PathLike) -> "MorseCode":
        """Construct from wave file

        - Read in wave file
        - Calculate signal envelope (smoothing of 0.1 seconds)
        - Apply squaring (threshold: 50% of max smoothed data value)

        Args:
            file (os.PathLike): path to input WAV file

        Returns:
            MorseCode: class instance, with 1D binary input data
        """
        sample_rate, wave = read_wave(file)
        window_size = int(0.01 * sample_rate)
        envelope = smoothed_power(wave, window_size)
        square_data = squared_signal(envelope)

        return cls(square_data)

    def decode(self) -> str:
        """Decode data

        Returns:
            str: Morse code content, in plain language

        Raises:
            UserWarning: dash/dot separation cannot be made unambiguosly
        """
        on_samples, off_samples = self._on_off_samples()
        dash_dot_chars = self._dash_dot_characters(on_samples)
        char_break_idx, word_space_idx = self._break_spaces(off_samples)
        morse_words = self._morse_words(dash_dot_chars, char_break_idx, word_space_idx)
        return self._translate(morse_words)

    @classmethod
    @property
    def morse_to_char(cls) -> dict[str, str]:
        """Morse to character dictionary

        Read mappings from morse.ini and store them to class variable. Later,
        return directly from this class variable.

        Returns:
            dict[str, str]: Mapping of morse character string to letter
        """
        if cls._morse_to_char is not None:
            return cls._morse_to_char

        config = ConfigParser()
        config.read(Path(__file__).parent / "morse.ini")
        chars = config["characters"]
        cls._morse_to_char = {chars[key]: key.upper() for key in chars}
        return cls._morse_to_char

    def _on_off_samples(self) -> tuple[np.ndarray, np.ndarray]:
        """Calculate signal ON/OFF durations

        Locate rising and falling edges in square wave at self.data. Calculate
        number of samples in each ON / OFF period.

        Returns:
            tuple[np.ndarray, np.ndarray]: on_samples, off_samples. Note that
                in addition to character and word spaces, off_samples also
                includes inter-character spaces.
        """
        if len(self.data) == 0:
            return np.array([], dtype="int"), np.array([], dtype="int")
        square_diff = np.diff(self.data)

        rising_idx = np.nonzero(square_diff == 1)[0]
        falling_idx = np.nonzero(square_diff == -1)[0]

        # Case: data starts with ON - it started one sample before index 0
        if falling_idx[0] < rising_idx[0]:
            rising_idx = np.insert(rising_idx, 0, -1)

        # Case: data ends with ON
        if rising_idx[-1] > falling_idx[-1]:
            falling_idx = np.insert(falling_idx, len(falling_idx), len(self.data) - 1)

        on_samples = falling_idx - rising_idx
        off_samples = rising_idx[1:] - falling_idx[: len(falling_idx) - 1]

        return on_samples, off_samples

    def _dash_dot_characters(self, on_samples: np.ndarray) -> np.ndarray:
        """Convert array of ON sample lengths to array of dashes and dots

        NOTE: It is expected, that the signal contains exactly two distinct
        lengths - those for a dash and for a dot. If the keying speed varies,
        or either character does not exist, then this method will fail.

        As a circumvention, 20 WPM is used as a guess

        Args:
            on_samples (np.ndarray): number of samples in each ON period in
                the signal. This comes from `MorseCode._on_off_samples`.

        Raises:
            UserWarning: if there are no distinct clusters (only dashes
                or dots in the input), and self.sample_rate is not set; thus
                no guess can be made on dash/dot.

        Returns:
            np.ndarray: array of dashes and dots, of object (string) type
        """
        if len(on_samples) == 0:
            return np.array([], dtype="str")
        n_clusters = min(2, len(on_samples))

        column_vec = on_samples.reshape(-1, 1)

        # Suppress ConvergenceWarning on too low distinct clusters; fix it later
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            clustering = sklearn.cluster.KMeans(
                n_clusters=n_clusters, random_state=0
            ).fit(column_vec)
        distinct_clusters = len(set(clustering.labels_))

        # It is not clear whether dash or dot -- use (20 wpm dot length) * 1.5 as limit
        if distinct_clusters == 1:
            if self.sample_rate is None:
                raise UserWarning("Cannot determine whether dash or dot")
            sys.stderr.write("WARNING: too little data, guessing based on 20 wpm")

            sample_length = clustering.cluster_centers_[0]
            is_dot = sample_length / (self.sample_rate * 60 / 1000) < 1.5
            dot_label = 0 if is_dot else 1
            dash_label = 1 if is_dot else 0
        else:
            cluster_sort_idx = np.argsort(
                clustering.cluster_centers_.flatten()
            ).tolist()
            dot_label = cluster_sort_idx.index(0)
            dash_label = cluster_sort_idx.index(1)

        dash_dot_map = {dot_label: ".", dash_label: "-"}
        dash_dot_characters = np.vectorize(dash_dot_map.get)(clustering.labels_)

        return dash_dot_characters

    @staticmethod
    def _break_spaces(off_samples: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
        """Convert array of OFF sample lengths to indices for char/word breaks

        NOTE: It is expected, that the signal contains exactly three distinct
        space lengths: inter-character space, character space and word space.
        If the keying speed varies, or word spaces do not exist, then this
        method will fail.

        Args:
            off_samples (np.ndarray): number of samples in each OFF period in
                the signal. This comes from `MorseCode._on_off_samples`.

        Returns:
            tuple[np.ndarray, np.ndarray]: indices for breaking dash/dot
                character array from `MorseCode._dash_dot_characters`. First
                array contains positions, where character breaks should be.
                Second array contains positions, where word spaces should be in
                the list of already resolved morse characters.
        """
        if len(off_samples) == 0:
            return np.array([], dtype="int"), np.array([], dtype="int")
        n_clusters = min(3, len(off_samples))

        column_vec = off_samples.reshape(-1, 1)

        # Suppress ConvergenceWarning on too low distinct clusters; fix it later
        with warnings.catch_warnings():
            warnings.simplefilter("ignore")
            clustering = sklearn.cluster.KMeans(
                n_clusters=n_clusters, random_state=0
            ).fit(column_vec)
        distinct_clusters = len(set(clustering.labels_))

        cluster_sort_idx = np.argsort(clustering.cluster_centers_.flatten()).tolist()

        # This index breaks dashes/dots into characters
        intra_space_label = cluster_sort_idx.index(0)
        char_break_idx = np.nonzero(clustering.labels_ != intra_space_label)[0] + 1

        char_or_word_space_arr = clustering.labels_[
            clustering.labels_ != intra_space_label
        ]

        # This index breaks character list into word lists
        if distinct_clusters == 3:
            word_space_label = cluster_sort_idx.index(2)
            word_space_idx = (
                np.nonzero(char_or_word_space_arr == word_space_label)[0] + 1
            )
        else:
            word_space_idx = np.array([], dtype="int")

        return char_break_idx, word_space_idx

    @staticmethod
    def _morse_words(
        raw_dash_dot: np.ndarray,
        char_break_idx: np.ndarray,
        word_space_idx: np.ndarray,
    ) -> list[list[str]]:
        """Convert character and space arrays to list of morse words

        Args:
            raw_dash_dot (np.ndarray): Numpy array of strings, contains
                '.' and '-' characters, as processed from self.data
            char_break_idx (np.ndarray): Index of locations in raw_dash_dot,
                where a character space or word space would exist. The array
                raw_dash_dot is first broken into characters with this index.
            word_space_idx (np.ndarray): Index for breaking character array
                into words. Contains locations of word spaces between natural
                language characters.

        Returns:
            list[list[str]]: Words in morse code. A single word is a list of
                dash-dot character combinations.
        """

        char_start_idx = [0] + (char_break_idx).tolist()
        char_end_idx = (char_break_idx).tolist() + [len(raw_dash_dot)]
        morse_characters = [
            "".join(raw_dash_dot[i:j].tolist())
            for i, j in zip(char_start_idx, char_end_idx)
        ]

        word_start_idx = [0] + (word_space_idx).tolist()
        word_end_idx = (word_space_idx).tolist() + [len(morse_characters)]

        return [morse_characters[i:j] for i, j in zip(word_start_idx, word_end_idx)]

    def _translate(self, morse_words: list[list[str]]) -> str:
        """Translate list of morse-coded words to string

        Args:
            morse_words (list[list[str]]): List of words, having list of characters.
                The characters are in morse-coded dash/dot form, e.g. '.--' for 'w'

        Returns:
            str: Message contained in input
        """
        char_dict = self.morse_to_char
        char_lists = [[char_dict.get(j, "") for j in i] for i in morse_words]
        return " ".join(["".join(word) for word in char_lists])
